.codegen/api.go.tmpl

```diff
@@ -162,70 +163,78 @@ func (a *{{.Service.Name}}API) {{.PascalName}}AndWait(ctx context.Context{{if .R
+	{{ if not .Request }}request := struct{}{}{{end}}
+	{{if eq .Pagination.Increment 1 -}}
+	request.{{.Pagination.Offset.PascalName}} = 1 // start iterating from the first page
+	{{end}}
+	getNextPage := func(ctx context.Context, req {{if .Request}}{{.Request.PascalName}}{{else}}struct{}{{end}}) (*{{ .Response.PascalName }}, error) {
+		ctx = useragent.InContext(ctx, "sdk-feature", "pagination")
+		return a.impl.{{.PascalName}}(ctx{{if .Request}}, req{{end}})
+	}
+	getItems := func(resp *{{ .Response.PascalName }}) []{{ template "type" .Pagination.Entity }} {
+		return {{if .Pagination.Results}}resp.{{.Pagination.Results.PascalName}}{{else}}resp{{end}}
+	}
+	{{if .Pagination.MultiRequest -}}
+	getNextReq := func(resp *{{ .Response.PascalName }}) *{{.Request.PascalName}} {
+		{{if .Pagination.Token -}}
+		if resp.{{.Pagination.Token.Bind.PascalName}} == "" {
+			return nil
+		}
+		request.{{.Pagination.Token.PollField.PascalName}} = resp.{{.Pagination.Token.Bind.PascalName}}
+		{{- else if eq .Path "/api/2.0/clusters/events" -}}
+		if resp.NextPage == nil {
+			return nil
+		}
+		request = *resp.NextPage
+		{{ else -}}
+		if len(getItems(resp)) == 0 {
+			return nil
+		}
+		{{ if eq .Pagination.Increment 1 -}}
+		request.{{.Pagination.Offset.PascalName}} = resp.{{.Pagination.Offset.PascalName}} + 1
+		{{- else -}}
+		request.{{.Pagination.Offset.PascalName}} = resp.{{.Pagination.Offset.PascalName}} + {{template "type" .Pagination.Offset.Entity}}(len(resp.{{.Pagination.Results.PascalName}}))
+		{{- end}}{{ end }}
+		return &request
+	}
+	{{- end}}
+	it = listing.NewIterator(
+		&request,
+		getNextPage,
+		getItems,
+		{{if .Pagination.MultiRequest}}getNextReq{{else}}nil{{end}})
+	{{ if .NeedsOffsetDedupe -}}
+	it = listing.NewDedupeIterator(
+		it,
+		func(item {{ template "type" .Pagination.Entity }}) {{ template "type" .IdentifierField.Entity }} {
+			return item{{ template "field-path" .IdFieldPath }}
+		})
+	{{- end }}
+	return it
+}
+
 func (a *{{.Service.Name}}API) {{.PascalName}}All(ctx context.Context{{if .Request}}, request {{.Request.PascalName}}{{end}}) ([]{{ template "type" .Pagination.Entity }}, error) {
-	{{if .Pagination.MultiRequest}}var results []{{.Pagination.Entity.PascalName}}
+	iter := a.{{.PascalName}}(ctx{{if .Request}}, request{{end}})
 	{{ if .Pagination.Limit -}}
+	var results []{{ template "type" .Pagination.Entity }}
 	var totalCount {{template "type" .Pagination.Limit.Entity}} = 0
-	{{ end -}}
-	ctx = useragent.InContext(ctx, "sdk-feature", "pagination")
-	{{if .NeedsOffsetDedupe -}}
-	// deduplicate items that may have been added during iteration
-	seen := map[{{template "type" .IdentifierField.Entity}}]bool{}
-	{{end}}{{if eq .Pagination.Increment 1 -}}
-	request.{{.Pagination.Offset.PascalName}} = 1 // start iterating from the first page
-	{{end}}for {
-		response, err := a.impl.{{.PascalName}}(ctx{{if .Request}}, request{{end}})
+	limit := request.{{.Pagination.Limit.PascalName}}
```

Comment:
This logic is a bit weird now. Limits should be applied at the iterator level, as these control the per-page limit.

Reply:
I'm trying to preserve the current behavior, do you think I did something wrong?

Eventually, I would like to remove all pagination parameters from request structures so that users don't misuse them or get confused by them anymore.

Reply:
No, it persist existing behavior. Just saying there is ambiguity in this approach; i.e. if I specify limit, does it apply to the number of elements per page, or the total number of elements returned.

Reply:
Agreed, it's very confusing. I think we made a change earlier to ensure that limit applied to the total returned number of elements when using the SDK, but that diverges from the REST API behavior (as well as from other SDK's behaviors). I do eventually want to remove these parameters from the list request structure, as well as all pagination-based parameters (for example, when using the SDK, why should a user be able to specify the token to use?).